sampledb/logic/schemas/validate.py

```diff
@@ -310,36 +310,67 @@ def _validate_quantity(instance: dict, schema: dict, path: typing.List[str]) ->
-        raise ValidationError('missing keys in schema: {}'.format(missing_keys), path)
+
+    #required_keys can no longer be statically defined, since either magnitude OR magnitude_in_base_units has to be given
+    #missing_keys = required_keys - schema_keys
+    #if missing_keys:
+    #    raise ValidationError('missing keys in schema: {}'.format(missing_keys), path)
+
     if instance['_type'] != 'quantity':
         raise ValidationError('expected _type "quantity"', path)
+    if 'units' not in instance:
+        instance['units'] = schema['units']
     if not isinstance(instance['units'], str):
         raise ValidationError('units must be str', path)
     if not units_are_valid(instance['units']):
         raise ValidationError('Invalid/Unknown units', path)
-    if not isinstance(instance['magnitude_in_base_units'], float) and not isinstance(instance['magnitude_in_base_units'], int):
-        raise ValidationError('magnitude_in_base_units must be float or int', path)
-    try:
-        quantity = datatypes.Quantity(instance['magnitude_in_base_units'], units=instance['units'])
-    except Exception:
-        raise ValidationError('Unable to create quantity', path)
-    if not isinstance(instance['dimensionality'], str):
-        raise ValidationError('dimensionality must be str', path)
     try:
         schema_quantity = datatypes.Quantity(1.0, units=schema['units'])
     except Exception:
         raise ValidationError('Unable to create schema quantity', path)
-    if quantity.dimensionality != schema_quantity.dimensionality:
+
+    quantity_magnitude = None
+    quantity_magnitude_in_base_units = None
+    if 'magnitude' in instance:
+        if not isinstance(instance['magnitude'], float) and not isinstance(instance['magnitude'], int):
+            raise ValidationError('magnitude must be float or int', path)
+        try:
+            quantity_magnitude = datatypes.Quantity(instance['magnitude'], units=instance['units'],
+                                                    already_in_base_units=False)
+        except Exception:
+            raise ValidationError('Unable to create quantity based on given magnitude', path)
+    if 'magnitude_in_base_units' in instance:
+        if not isinstance(instance['magnitude_in_base_units'], float) and not isinstance(instance['magnitude_in_base_units'], int):
+            raise ValidationError('magnitude_in_base_units must be float or int', path)
+        try:
+            quantity_magnitude_in_base_units = datatypes.Quantity(instance['magnitude_in_base_units'], units=instance['units'],
+                                                                  already_in_base_units=True)
+        except Exception:
+            raise ValidationError('Unable to create quantity based on given magnitude_in_base_units', path)
+
+    if quantity_magnitude is not None and quantity_magnitude_in_base_units is not None \
+            and quantity_magnitude.magnitude != quantity_magnitude_in_base_units.magnitude:
```

Comment:
Will the values calculated by pint always fulfill strict float equality? Otherwise, something like `math.isclose` should be used here so that values returned by the API will also be valid as input to the API.

Reply:
Well i think that really is the only way to do it here. I am checking if the given magnitude_in_base_units via pint converts into the exact same magnitude as given. If not the script (web interface or API request) definitely contains an error. In my mind the value in the schema defined unit always has to stay exaclty the same. It is essentially the defined frame of reference.
The magnitude_in_base_units on the other hand might be very slightly different, depending on how far apart unit and base_unit are (not sure about this though). The slight differences in the magnitude_in_base_units might still convert into the exact same magnitude but cause tiny diffences in other units that are ok and also should still be essentially stored in the database (which is now possible since i dont just overwrite the instance value anymore :) ). But thats why i changed it from comparing whole pint objects to just comparing the magnitude for equality.
For the sake of data integrity i just think that there should be no tolerance at all via "isclose" on the magnitude values and there always exists a corresponding base unit value that pint converts into exactly the magnitude one wants.

Reply:
Usually this will be the case, but for some units or magnitudes this will not be possible, e.g. the conversion from degrees Fahrenheit to Kelvin and back will result in a floating point error, about 2e-14˚F for an input of 1˚K.

When working with floating point numbers, this seems like normal and expected behavior to me, and I don't think exact equality should be expected while storing values as floating point numbers. Otherwise, we might already have issues whenever a user wishes to store a value such as 0.1 which cannot be represented exactly.

Reply:
You are absolutely right, i feel kind of stupid now, because somehow i thought there always has to be a base_unit value to create the exact same magnitude, but thats clearly wrong. I just couldn't think of an example. 
I am now admitting that you can't set it up i a way to always ensure equality after the pint conversion from 2 different endpoints, so i changed the comparison to math.round. 

Additionally i prevented the magnitude to be recalculated in the from_json method if its already part of the json data while being backwards compatible to old stored json that only know the base_units value. No need for a potential recalculation inaccuracy there considering we now store and preserve the originally entered magnitude value as well.